### PR TITLE
Integration of SSP333 time-stepper

### LINKED ISSUE #1341: Add support for `SSP333` time stepper
Benefits:
The new SSP time stepper, including its integration with limiters, will be tested in ClimaAtmos.

Risks:
Lack of simpler, incremental examples in the upstream repo. There is not currently a working spherical 3D simple example that uses the new SSP algorithm yet in ClimaTimeSteppers (only a 2D one), and there are no examples with multiple tracers (that we'd need to account for in ClimaAtmos). 
Documentation of the new interfaces is lacking and scattered.

### PR DESCRIPTION
**Pull request by valeriabarra**
## Purpose 

The purpose of this PR is to add integration of the new SSP time-stepper. 

Closes #1341 

## Content
- [x] Add new `horizontal_limiter_tendency!`
- [x] Move limiters definition inside `limiters_func!`
- [x] Pass `horizontal_limiter_tendency!` as `T_lim!` and `limiters_func!` as `lim!` to the `ClimaODEFunction`
- [x] Define a new `hyperdiffusion_tracers_tendency!` that strips out the tracers components from the original `hyperdiffusion_tendency!`
- [x] Add relevant regular and longrun CI jobs for testing (moist baroclinic wave, moist HS, and moist aquaplanet with `clearsky` radiation) 
- [x] Rename existing CI jobs:
    - `baroclinic wave (ρe) equilmoist` -> `no lim ARS baroclinic wave (ρe) equilmoist`
    - `baroclinic wave (ρe) equilmoist check conservation float64` -> `no lim ARS baroclinic wave (ρe) equilmoist check conservation float64`
    - `held suarez (ρe) equilmoist hightop sponge` -> `no lim ARS held suarez (ρe) equilmoist hightop sponge`
    -  `MPI aquaplanet (ρe) equilmoist clearsky radiation` -> `MPI no lim ARS aquaplanet (ρe) equilmoist clearsky radiation`.

Review checklist

I have:
- followed the codebase contribution guide: https://clima.github.io/ClimateMachine.jl/latest/Contributing/
- followed the style guide: https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/
- followed the documentation policy: https://github.com/CliMA/policies/wiki/Documentation-Policy
- checked that this PR does not duplicate an open PR.

In the Content, I have included 
- relevant unit tests, and integration tests, 
- appropriate docstrings on all functions, structs, and modules, and included relevant documentation.

- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

**Comment by szy21** [comment]
The artifacts look good to me, except for zonal wind and kinetic energy spectrum in baroclinic waves. For example see `bw-u1500-day10.png` and `bw-n_integrated_ke_spectrum1500-day10` [here](https://buildkite.com/clima/climaatmos-ci/builds/7365#0186777b-f2a8-47b5-bf20-2c6a8f3ef45e) and compare to current main.

**Comment by valeriabarra** [comment]
> The artifacts look good to me, except for zonal wind and kinetic energy spectrum in baroclinic waves. For example see `bw-u1500-day10.png` and `bw-n_integrated_ke_spectrum1500-day10` [here](https://buildkite.com/clima/climaatmos-ci/builds/7365#0186777b-f284-4b5e-915e-bb3e8222c53e) and compare to current main.

Here are the zonal wind and pressure plots from this branch as compared to the main branch (which does not have limiters hooked-in, since some time in December) plotted on the same color map scale, as discussed today.

- This branch, zonal wind, day 10:
bw-u1500-day10_branch

- `main` branch, zonal wind, day 10:
bw-u1500-day10_main


- This branch, pressure, day 10:
bw-pressure-day10_branch

- `main` branch, pressure, day 10:
bw-pressure-day10_main

- As an aside, this is the `qt` plot at day 10 from this branch, showing no negative values:
bw-qt-day10_branch

- And this is the corresponding `qt` plot on `main`, showing negative values:
bw-qt-day10_main

**Comment by szy21** [comment]
Thanks @valeriabarra! I think it's probably fine, as some small change in qt affects pressure, which affects u. I would expect much smaller changes in the steady state of moist held suarez. But it would be good if @tapios can double check.

**Comment by tapios** [comment]
LGTM. But I'd also like to see the statistically steady state of a moist simulation.

**Comment by tapios** [comment]
Is there a substantial difference in time step/time to solution with SSP IMEX vs ARS? And do we need multiple Newton iterations for this to be stable?

**Comment by valeriabarra** [comment]
The time to solution is shorter for SSP compared to ARS. As it can be seen in this [build ](https://buildkite.com/clima/climaatmos-ci/builds/7365) the moist baroclinic wave w/ SSP took 22 minutes, as opposed to 27 minutes w/ ARS scheme; and the moist high-top HS took 34 minutes w/ the SSP, as opposed to 40 minutes for the ARS scheme. For these simulations, I did not have to adjust the number of Newton's iterations, so it is still equal to 1, as its default value.

**Comment by tapios** [comment]
That's nice to hear! (It contradicts Gardner et al. Maybe they had some issues with their implementation.)

**Comment by szy21** [comment]
I'm not sure if the time to solution is shorter for SSP, as you are using the same dt (300secs) for SSP and ARS now. I don't know whether this is the maximum timestep, but we should be able to use a larger dt (e.g. 400secs) for ARS. It probably makes sense that the maximum timestep is larger for ARS as it has more stages. But it seems at least there is no substantial difference in time-to-solution.

**Comment by valeriabarra** [comment]
Yes, the reduction of the time step was just a test while I was debugging. I will put it back now that's been cleared

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -120,64 +120,76 @@ steps:
...
 
       - label: ":computer: baroclinic wave (ρe) equilmoist"
-        command: 
+        command:
           - julia --color=yes --project=examples examples/hybrid/driver.jl --moist equil --precip_model 0M --job_id sphere_baroclinic_wave_rhoe_equilmoist --regression_test true --dt 450secs --dt_save_to_disk 5days
           - julia --color=yes --project=examples post_processing/remap/remap_pipeline.jl --data_dir sphere_baroclinic_wave_rhoe_equilmoist --out_dir sphere_baroclinic_wave_rhoe_equilmoist
         artifact_paths: "sphere_baroclinic_wave_rhoe_equilmoist/*"
 
+      - label: ":computer: SSP baroclinic wave (ρe_tot) equilmoist"
+        command: "julia --color=yes --project=examples examples/hybrid/driver.jl --moist equil --precip_model 0M --job_id sphere_ssp_baroclinic_wave_rhoe_equilmoist --regression_test true --dt 450secs --ode_algo SSP333"
```
Comment (szy21) [inline .buildkite/pipeline.yml:149]:
```suggestion
        command:
          - julia --color=yes --project=examples examples/hybrid/driver.jl --moist equil --precip_model 0M --job_id sphere_ssp_baroclinic_wave_rhoe_equilmoist --regression_test true --dt 450secs --ode_algo SSP333
          - julia --color=yes --project=examples post_processing/remap/remap_pipeline.jl --data_dir sphere_ssp_baroclinic_wave_rhoe_equilmoist --out_dir sphere_ssp_baroclinic_wave_rhoe_equilmoist
```

Comment (szy21) [reply .buildkite/pipeline.yml:149]: (if you would like to have remapped files)

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -120,64 +120,76 @@ steps:
...
           - julia --color=yes --project=examples post_processing/remap/remap_pipeline.jl --data_dir sphere_held_suarez_rhoe_equilmoist_hightop_sponge --out_dir sphere_held_suarez_rhoe_equilmoist_hightop_sponge
         artifact_paths: "sphere_held_suarez_rhoe_equilmoist_hightop_sponge/*"
 
+      - label: ":computer: SSP held suarez (ρe) equilmoist hightop sponge"
+        command: "julia --color=yes --project=examples examples/hybrid/driver.jl --z_max 45e3 --z_elem 25 --dz_bottom 300 --vert_diff true --surface_scheme bulk --moist equil --forcing held_suarez --precip_model 0M --rayleigh_sponge true --alpha_rayleigh_uh 0 --viscous_sponge true --zd_rayleigh 30e3 --zd_viscous 30e3 --job_id sphere_ssp_held_suarez_rhoe_equilmoist_hightop_sponge --dt 450secs --t_end 5days --dt_save_to_disk 5days --regression_test true --ode_algo SSP333"
```
Comment (szy21) [inline .buildkite/pipeline.yml:179]: same as above, you can add a command for remapping

File: .buildkite/longruns/pipeline.yml [hunk 3]
```diff
@@ -53,6 +53,24 @@ steps:
         agents:
           slurm_ntasks: 32
 
+      - label: ":computer: baroclinic wave (ρe_tot) high resolution topography (dcmip)"
```
Comment (szy21) [inline .buildkite/longruns/pipeline.yml:56]: This job has been moved to "experimental", so we don't need it here

File: .buildkite/longruns/pipeline.yml [hunk 4]
```diff
@@ -63,8 +81,38 @@ steps:
...
+        command:
+          - "mpiexec julia --project=examples examples/hybrid/driver.jl --moist equil --precip_model 0M --z_elem 45 --dz_bottom 30 --h_elem 16 --kappa_4 1e16 --dt 400secs --t_end 100days --dt_save_to_disk 10days --job_id longrun_ssp_bw_rhoe_equil_highres --ode_algo SSP333"
+          - julia --color=yes --project=examples post_processing/remap/remap_pipeline.jl --data_dir longrun_ssp_bw_rhoe_equil_highres --out_dir longrun_ssp_bw_rhoe_equil_highres
+        artifact_paths: "longrun_ssp_bw_rhoe_equil_highres/*"
+        env:
+          CLIMACORE_DISTRIBUTED: "MPI"
+        agents:
+          slurm_ntasks: 32
+
+      - label: ":computer: baroclinic wave (ρe_tot) equilmoist high resolution topography (dcmip)"
```
Comment (szy21) [inline .buildkite/longruns/pipeline.yml:104]: Same as above, this job has been moved to "experimental", so we don't need it here.

File: .buildkite/longruns/pipeline.yml [hunk 5]
```diff
@@ -73,8 +121,16 @@ steps:
         agents:
           slurm_ntasks: 32
 
+      - label: ":computer: no_lim held suarez (ρe_tot) equilmoist high resolution"
+        command: "mpiexec julia --project=examples examples/hybrid/driver.jl --forcing held_suarez --moist equil --vert_diff true --surface_scheme bulk --precip_model 0M --z_elem 45 --dz_bottom 30 --h_elem 16 --bubble true --kappa_4 1e16 --dt 150secs --t_end 400days --dt_save_to_disk 10days --job_id longrun_no_lim_hs_rhoe_equil_highres --apply-limiter false"
```
Comment (szy21) [inline .buildkite/longruns/pipeline.yml:125]:
```suggestion
        command: 
          - mpiexec julia --project=examples examples/hybrid/driver.jl --forcing held_suarez --moist equil --vert_diff true --surface_scheme bulk --precip_model 0M --z_elem 45 --dz_bottom 30 --h_elem 16 --bubble true --kappa_4 1e16 --dt 150secs --t_end 400days --dt_save_to_disk 10days --job_id longrun_no_lim_hs_rhoe_equil_highres --apply-limiter false"
          - julia --color=yes --project=examples post_processing/remap/remap_pipeline.jl --data_dir longrun_no_lim_hs_rhoe_equil_highres --out_dir longrun_no_lim_hs_rhoe_equil_highres
```

File: .buildkite/longruns/pipeline.yml [hunk 6]
```diff
@@ -83,8 +139,40 @@ steps:
         agents:
           slurm_ntasks: 64
 
+     - label: ":computer: no lim SSP held suarez (ρe_tot) equilmoist high resolution"
+        command: "mpiexec julia --project=examples examples/hybrid/driver.jl --forcing held_suarez --moist equil --vert_diff true --surface_scheme bulk --precip_model 0M --z_elem 45 --dz_bottom 30 --h_elem 16 --bubble true --kappa_4 1e16 --dt 150secs --t_end 400days --dt_save_to_disk 10days --job_id longrun_no_lim_ssp_hs_rhoe_equil_highres --ode_algo SSP333 --apply_limiter false"
```
Comment (szy21) [inline .buildkite/longruns/pipeline.yml:143]: Same as above, you can add a command for remapping.

File: .buildkite/longruns/pipeline.yml [hunk 7]
```diff
@@ -83,8 +139,40 @@ steps:
...
+        artifact_paths: "longrun_no_lim_ssp_hs_rhoe_equil_highres/*"
+        env:
+          CLIMACORE_DISTRIBUTED: "MPI"
+        agents:
+          slurm_ntasks: 64
+
+      - label: ":computer: SSP held suarez (ρe_tot) equilmoist high resolution"
+        command: "mpiexec julia --project=examples examples/hybrid/driver.jl --forcing held_suarez --moist equil --vert_diff true --surface_scheme bulk --precip_model 0M --z_elem 45 --dz_bottom 30 --h_elem 16 --bubble true --kappa_4 1e16 --dt 150secs --t_end 400days --dt_save_to_disk 10days --job_id longrun_ssp_hs_rhoe_equil_highres --ode_algo SSP333"
```
Comment (szy21) [inline .buildkite/longruns/pipeline.yml:151]: Same as above, you can add a command for remapping.

File: .buildkite/longruns/pipeline.yml [hunk 8]
```diff
@@ -83,8 +139,40 @@ steps:
...
+        artifact_paths: "longrun_ssp_hs_rhoe_equil_highres/*"
+        env:
+          CLIMACORE_DISTRIBUTED: "MPI"
+        agents:
+          slurm_ntasks: 64
+
+      - label: ":computer: held suarez (ρe_tot) equilmoist high resolution topography (dcmip)"
```
Comment (szy21) [inline .buildkite/longruns/pipeline.yml:158]: This job has been moved to "experimental", so we don't need it here.

File: .buildkite/longruns/pipeline.yml [hunk 9]
```diff
@@ -83,8 +139,40 @@ steps:
...
+        command: "mpiexec julia --project=examples examples/hybrid/driver.jl --forcing held_suarez --moist equil --vert_diff true --topography DCMIP200 --surface_scheme bulk --precip_model 0M --z_elem 45 --dz_bottom 30 --h_elem 16 --bubble true --kappa_4 1e16 --dt 150secs --t_end 400days --dt_save_to_disk 10days --job_id longrun_hs_rhoe_equil_highres_topography --rayleigh_sponge true --viscous_sponge true"
+        artifact_paths: "longrun_hs_rhoe_equil_highres_topography/*"
+        env:
+          CLIMACORE_DISTRIBUTED: "MPI"
+        agents:
+          slurm_ntasks: 64
+
+      - label: ":computer: no lim aquaplanet (ρe_tot) equilmoist high resolution clearsky radiation Float64"
+        command: "mpiexec julia --project=examples examples/hybrid/driver.jl --vert_diff true --surface_scheme monin_obukhov --moist equil --rad clearsky --precip_model 0M --z_elem 45 --dz_bottom 30 --h_elem 16 --bubble true --kappa_4 1e16 --rayleigh_sponge true --alpha_rayleigh_uh 0 --dt 150secs --t_end 400days --fps 30 --job_id longrun_no_lim_aquaplanet_rhoe_equil_highres_clearsky_ft64 --dt_save_to_sol 10days --dt_save_to_disk 10days --FLOAT_TYPE Float64 --apply_limiter false"
```
Comment (szy21) [inline .buildkite/longruns/pipeline.yml:167]: Same as above, you can add a command for remapping.

File: .buildkite/longruns/pipeline.yml [hunk 10]
```diff
@@ -93,8 +181,24 @@ steps:
         agents:
           slurm_ntasks: 64
 
+      - label: ":computer: no lim SSP aquaplanet (ρe_tot) equilmoist high resolution clearsky radiation Float64"
+        command: "mpiexec julia --project=examples examples/hybrid/driver.jl --vert_diff true --surface_scheme monin_obukhov --moist equil --rad clearsky --precip_model 0M --z_elem 45 --dz_bottom 30 --h_elem 16 --bubble true --kappa_4 1e16 --rayleigh_sponge true --alpha_rayleigh_uh 0 --dt 150secs --t_end 400days --fps 30 --job_id longrun_no_lim_ssp_aquaplanet_rhoe_equil_highres_clearsky_ft64 --dt_save_to_sol 10days --dt_save_to_disk 10days --FLOAT_TYPE Float64 --ode_algo SSP333 --apply_limiter false"
```
Comment (szy21) [inline .buildkite/longruns/pipeline.yml:185]: Same as above, you can add a command for remapping.

File: .buildkite/longruns/pipeline.yml [hunk 11]
```diff
@@ -93,8 +181,24 @@ steps:
...
+        artifact_paths: "longrun_no_lim_ssp_aquaplanet_rhoe_equil_highres_clearsky_ft64/*"
+        env:
+          CLIMACORE_DISTRIBUTED: "MPI"
+        agents:
+          slurm_ntasks: 64
+
+      - label: ":computer: SSP aquaplanet (ρe_tot) equilmoist high resolution clearsky radiation Float64"
+        command: "mpiexec julia --project=examples examples/hybrid/driver.jl --vert_diff true --surface_scheme monin_obukhov --moist equil --rad clearsky --precip_model 0M --z_elem 45 --dz_bottom 30 --h_elem 16 --bubble true --kappa_4 1e16 --rayleigh_sponge true --alpha_rayleigh_uh 0 --dt 150secs --t_end 400days --fps 30 --job_id longrun_ssp_aquaplanet_rhoe_equil_highres_clearsky_ft64 --dt_save_to_sol 10days --dt_save_to_disk 10days --FLOAT_TYPE Float64 --ode_algo SSP333"
```
Comment (szy21) [inline .buildkite/longruns/pipeline.yml:193]: Same as above, you can add a command for remapping.

**Review by szy21 — APPROVED** [review]

**Comment by bischtob** [comment]
Notes: Longruns are going unstable at 1 Newton iteration. For now, works only when Newton iterations are increased. @tapios mentioned that one Newton iteration should be sufficient. Gardner paper however has more than 1 Newton iteration.

**Comment by valeriabarra** [comment]
These are longrun moist baroclinic wave results:
- This branch, mean `u` against sigma:
   
mean_u_sigma_ssp_branch

- As opposed to `main` branch, mean `u` against sigma:
   
mean_u_sigma_main

 
I will report on the longrun moist HS run as soon as I have it.

**Comment by valeriabarra** [comment]
I would like to merge this and if issues arise in other longruns, I'll open a separate issue (as originally planned in the SDI).
 @szy21 what do you think? Can we go ahead and squash/merge?

**Comment by szy21** [comment]
> I would like to merge this and if issues arise in other longruns, I'll open a separate issue (as originally planned in the SDI). @szy21 what do you think? Can we go ahead and squash/merge?

Yes, that sounds good to me. We can merge this.

**Comment by trontrytel** [comment]
The EDMF SCM Bomex JFNK IMEX case is slightly changed (very slightly in the final profiles and a little more in the initial time evolution). We looked at the tendencies between this PR and main branch and did not see any differences. I run this branch with smaller dt and the time evolution has returned back to normal.I went over the code and see no reason why SCM cases should be even affected.

EDMF JFNK IMEX is an experimental case that we are still working on fixing/improving/debuging. I'm fine with merging this PR. Hopefully we will make the EDMF IMEX more stable in the upcoming days.

File: .buildkite/pipeline.yml [hunk 12]
```diff
@@ -124,64 +124,80 @@ steps:
...
+          - julia --color=yes --project=examples examples/hybrid/driver.jl --moist equil --precip_model 0M --job_id sphere_baroclinic_wave_rhoe_equilmoist --regression_test true --dt 450secs --dt_save_to_disk 5days --apply_limiter false
           - julia --color=yes --project=examples post_processing/remap/remap_pipeline.jl --data_dir sphere_baroclinic_wave_rhoe_equilmoist --out_dir sphere_baroclinic_wave_rhoe_equilmoist
         artifact_paths: "sphere_baroclinic_wave_rhoe_equilmoist/*"
 
-      - label: ":computer: baroclinic wave (ρe) equilmoist check conservation float64"
-        command: "julia --color=yes --project=examples examples/hybrid/driver.jl --moist equil --rad clearsky --job_id sphere_baroclinic_wave_rhoe_equilmoist_conservation_ft64 --check_conservation true --FLOAT_TYPE Float64 --bubble true --dt 450secs --dt_save_to_disk 5days"
+      - label: ":computer: SSP baroclinic wave (ρe_tot) equilmoist"
+        command:
+          - "julia --color=yes --project=examples examples/hybrid/driver.jl --moist equil --precip_model 0M --job_id sphere_ssp_baroclinic_wave_rhoe_equilmoist --regression_test true --dt 450secs --ode_algo SSP333 --regression_test true"
```
Comment (szy21) [inline .buildkite/pipeline.yml:154]:
```suggestion
          - "julia --color=yes --project=examples examples/hybrid/driver.jl --moist equil --precip_model 0M --job_id sphere_ssp_baroclinic_wave_rhoe_equilmoist --regression_test true --dt 450secs --ode_algo SSP333 --regression_test true --dt_save_to_disk 5days"
```
(You called the remap but did not save any hdf5 files, and that's why the buildkite job failed I think.)

Comment (valeriabarra) [reply .buildkite/pipeline.yml]: Oh, thank you for spotting that!

Comment (valeriabarra) [reply .buildkite/pipeline.yml]: Sorry for the noise of notifications!

**Pull request merged**